library/cc/stream_callbacks.h

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <cstddef>
+#include <optional>
+#include <vector>
+
+#include "envoy_error.h"
+#include "executor.h"
+#include "response_headers.h"
+#include "response_trailers.h"
+
+using OnHeadersCallback = std::function<void(ResponseHeaders headers, bool end_stream)>;
+using OnDataCallback = std::function<void(std::vector<std::byte> data, bool end_stream)>;
+using OnTrailersCallback = std::function<void(ResponseTrailers trailers)>;
+using OnCancelCallback = std::function<void()>;
+using OnErrorCallback = std::function<void(EnvoyError error)>;
+
+struct StreamCallbacks {
+  std::optional<OnHeadersCallback> on_headers;
+  std::optional<OnDataCallback> on_data;
+  std::optional<OnTrailersCallback> on_trailers;
+  std::optional<OnCancelCallback> on_cancel;
+  std::optional<OnErrorCallback> on_error;
+};
+
+class EnvoyHttpCallbacksAdapter {
+public:
+  EnvoyHttpCallbacksAdapter(Executor executor, StreamCallbacks callbacks);
+
+private:
+  // TODO(crockeo): figure out lifecycle management here
+  Executor& executor_;
```

Comment:
i've been futzing around in rust-land where i get to feel comfortable about reference lifetimes. this isn't really _safe_ per se, but its lifetime will be articulated in a [pybind11::keep_alive](https://pybind11.readthedocs.io/en/stable/advanced/functions.html#keep-alive) on the python obj definition

Reply:
You could use std::weak_ptr for a stronger guarantee. Or even just a shared_ptr to ensure its lifetime.

Reply:
the best case scenario for this is that someone in python land creates and then immediately moves ownership of the Executor into the C++ codebase. i'm not sure that we can enforce move semantics in python, though, so that may be a moot attempt.

i'd say let's use `std::unique_ptr` for now to represent the ideal world and then we can backtrack into whatever ends up working (probably a `std::weak_ptr` or `std::shared_ptr`

Reply:
>In other words, there is nothing special that needs to be done. While returning unique pointers in this way is allowed, it is illegal to use them as function arguments. For instance, the following function signature cannot be processed by pybind11.

[from pybind11 docs](https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html#std-unique-ptr)

looks like we're going to have to use `shared_ptr`s